packages/paymaster/src/BiconomyPaymaster.ts

```diff
@@ -1,49 +1,241 @@
-import { sendRequest, HttpMethod } from '@biconomy/common'
+import { Logger, sendRequest, HttpMethod } from '@biconomy/common'
 import { resolveProperties } from '@ethersproject/properties'
-import { UserOperation } from '@biconomy/core-types'
-import { IPaymaster } from 'interfaces/IPaymaster'
-import { PaymasterAndDataResponse, PaymasterConfig } from './utils/Types'
+import { UserOperation, Transaction } from '@biconomy/core-types'
+import { Provider } from '@ethersproject/abstract-provider'
+import {
+  PaymasterFeeQuote,
+  PaymasterConfig,
+  BiconomyTokenPaymasterFeeQuoteResponse,
+  FeeQuotesOrDataDto,
+  SponsorUserOperationDto,
+  PaymasterServiceSuccessResponse,
+  BiconomyTokenPaymasterRequest
+} from './utils/Types'
+import { BigNumberish, BigNumber, ethers } from 'ethers'
+// TODO
+import { ERC20_ABI } from './constants' // temporary
+import { IHybridPaymaster } from './interfaces/IHybridPaymaster'
 
-export class BiconomyPaymaster implements IPaymaster {
+// Hybrid - Generic Gas abstraction paymaster
+/**
+ *
+ */
+export class BiconomyPaymaster implements IHybridPaymaster {
   constructor(readonly paymasterConfig: PaymasterConfig) {}
+
+  // TODO // WIP
+  // Note: maybe rename to createTokenApprovalTransaction
+  async createTokenApprovalRequest(
+    tokenPaymasterRequest: BiconomyTokenPaymasterRequest,
+    provider: Provider
+  ): Promise<Transaction> {
+    // Note: should also check in caller if the approval is already given
+
+    const feeTokenAddress: string = tokenPaymasterRequest.feeQuote.tokenAddress
+    Logger.log('erc20 fee token address ', feeTokenAddress)
+
+    // TODO: For some tokens we may need to set allowance to 0 first so that would return batch of transactions
+    const requiredApproval: number = Math.floor(
+      tokenPaymasterRequest.feeQuote.maxGasFee *
+        Math.pow(10, tokenPaymasterRequest.feeQuote.decimal)
+    )
+    Logger.log('required approval for erc20 token ', requiredApproval)
+
+    const spender = tokenPaymasterRequest.spender
+    // fallback to fetch from member
+    // this.paymasterAddress ? this.paymasterAddress : await this.getPaymasterAddress()
+    // Might maintain two fields for token and verifying paymaster addresses
+
+    const erc20 = new ethers.Contract(feeTokenAddress, ERC20_ABI, provider)
+
+    return {
+      to: erc20.address,
+      value: ethers.BigNumber.from(0),
+      data: erc20.interface.encodeFunctionData('approve', [spender, requiredApproval.toString()])
```

Comment:
currently it's using this

```const requiredApproval: number = Math.floor(
      tokenPaymasterRequest.feeQuote.maxGasFee *
        Math.pow(10, tokenPaymasterRequest.feeQuote.decimal)
    )```
    
since maxGasFee and decimal received from feeQuote are numbers. like we ran into error parsing exchangeRate as BigNumber, these could be similarly large values

Reply:
maxGasFee can be decimals

Reply:
```
/*requiredApproval = ethers.BigNumber.from(
        tokenPaymasterRequest.feeQuote.maxGasFee.toString()
      ).mul(
        ethers.BigNumber.from(10).pow(
          ethers.BigNumber.from(tokenPaymasterRequest.feeQuote.decimal.toString())
        )
      )*/
```

this might cause issues